Add new API for querying chunks

VinGarcia opened the pull request:
## Intro

This is a proposal of change on the ksql API. This change will be 100% retro compatible since it will just add a new function, and deprecate the existing one (which implies removing it from the docs and adding a `// depreacated` comment on it).

The new function is just as safe as the old one but requires an extra error check, which is not ideal, but probably preferable since it still looks better with the error check than the old way of doing it looked like.

## Short description of the QueryChunks feature

The QueryChunks feature is a necessary feature that allows the user to make queries that return very big amounts of data. What this feature does is that instead of loading all the data at once in memory, it loads chunks of the response at a time, which circumvents the problem of overloading the computer's available memory.

## Proposal

This PR is an experimental idea of changing the current API for querying chunks, which looks like this:

```golang
	err = db.QueryChunks(ctx, ksql.ChunkParser{
		Query:     "SELECT * FROM users WHERE type = $1",
		Params:    []interface{}{usersType},
		ChunkSize: 100,
		ForEachChunk: func(users []User) error {
			err := sendUsersSomewhere(users)
			if err != nil {
				// This will abort the QueryChunks loop and return this error
				return err
			}
			return nil
		},
	})
	if err != nil {
		panic(err.Error())
	}
```

To something that looks more natural and easier to read like this:

```golang
	var chunks ksql.Chunks
	err := ksql.Query(ctx, &chunks, `SELECT * FROM users WHERE type = $1`, usersType)
	if err != nil {
		panic(err.Error())
	}

	err := chunks.ForEach(100, func(users []User) error {
		err := sendUsersSomewhere(users)
		if err != nil {
			// This will abort the ForEach loop and return this error
			return err
		}
		return nil
	})
	if err != nil {
		panic(err.Error())
	}
```

In this scenario the `ksql.Query()` function will have a special behavior if the record is passed as an argument is of type `ksql.Chunks`.

This special behavior is to instantiate a new `ksql.chunks{}` struct that has a `.ForEach()` method that behaves just like the existing `ksql.QueryChunks()` function, but requiring less arguments (namely only the `chunkSize` and the for-each callback).

## Possible advantages of this syntax for the future

This syntax could allow us to setup some interesting features such as prepared statements in the future. Since the Query step is done first we could in that step prepare the statement.

This technique could even extend to other types of queries too.

However, if we end up finding a way of implementing a system of creating prepared statements automatically internally this idea might not need to be implemented after all.

cvelosomg commented:
Test

derkan commented:
Great idea, what about this syntax?

```go
err := ksql.QueryFor(ctx, 100, `SELECT * FROM users WHERE type = $1`, usersType).Each(
	func(users []User) error {
		if err := sendUsersSomewhere(users); err != nil {
			return err
		}
		return nil
	})
if err != nil {
	panic(err.Error())
}
```

VinGarcia commented:
Hey, @derkan first sorry for taking so long to answer, I missed the email notification.

But yeah I am thinking about something a lot like that, but I am struggling with a few issues that will most likely make me postpone this change to a v2 version:

So initially I decided I would not have currying in this library as a design principle. The reason was that whenever I found an open-source library that had currying this automatically made it much harder to write mocks for that library, and I wanted mocking to be easy even if you used some other mocking library like GoMock.

So, I actually had a few principles I wanted to fit in this library, and the current API is pretty much the result of the best trade-off I was able to find with all these principles:

- Simplicity
- Easy to test and mock
- Handling errors only once
- and a few others

So the proposal of this PR would be beneficial in the simplicity front, but for that I would not want to add new functions to API or the gain I got with the change would be lost.

And at the same time I am losing the principle of returning errors once for each operation which is also something I wanted to keep.

So ok, I am making a big problem out of something simple, but I don't have any hurry to add new features, so I will go easy on this one.

One other piece of information that might be worth sharing is that I am actually having some good ideas that could improve the simplicity a lot if I just dropped this "no currying" principle and rewrote part of the API. And of course, I won't do this in the v1 version of this library because I want it to be very stable. But I could make this on a V2 version, the syntax you proposed was exactly what I was considering.

I could return something like a `ksql.Query` type from the `.Query()` function, and then this Query type would have 4 methods:

1. `ForEach()`
2. `GetOne()`
3. `GetAll()`
4. `Prepare()` (this one would prepare the statement for performance reasons)

This would allow me to also reduce the number of mocking tools on the `ksqltest` package and reduce the number of functions in the `Provider` API, so I am thinking this might be a change worth doing on V2.

However, after evaluating this change on this version I noticed that I will have to make more changes to the `ksqltest` package than I expected which will put to waste the gains I got in the simplicity front, at least on V1, so I am thinking about closing this PR and focusing on finishing the TODO list I have on the README. In the future I will work on this new API for the V2 version.

@jadson-medeiros thank you for the review, but for now I will close this PR.
